src/irr/asset/CImageLoaderOpenEXR.cpp

```diff
@@ -41,17 +41,69 @@ namespace irr
 
 			SContext ctx;
 
-			readVersionField(_file, ctx);
-			readHeader(fileName, ctx);
+			// readVersionField(_file, ctx);
+			// readHeader(fileName, ctx);
 
 			Array2D<Rgba> pixels;
 			int width;
 			int height;
 
 			readRgba(fileName, pixels, width, height);
 
-			// TODO put data into buffer, consider OpenEXR 2.0 supported saving layouts 
-			return SAssetBundle{};
+			constexpr uint32_t MAX_PITCH_ALIGNMENT = 8u;									   // OpenGL cannot transfer rows with arbitrary padding
+			auto calcPitchInBlocks = [](uint32_t width, uint32_t blockByteSize) -> uint32_t	   // try with largest alignment first
+			{
+				auto rowByteSize = width * blockByteSize;
+				for (uint32_t _alignment = MAX_PITCH_ALIGNMENT; _alignment > 1u; _alignment >>= 1u)
+				{
+					auto paddedSize = core::alignUp(rowByteSize, _alignment);
+					if (paddedSize % blockByteSize)
+						continue;
+					return paddedSize / blockByteSize;
+				}
+				return width;
+			};
+
+			ICPUImage::SCreationParams params;
+			params.format = EF_R16G16B16A16_SFLOAT;
+			params.type = ICPUImage::ET_2D;;
+			params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0u);
+			params.samples = ICPUImage::ESCF_1_BIT;
+			params.extent.width = width;
+			params.extent.height = height;
+			params.extent.depth = 1u;
+			params.mipLevels = 1u;
+			params.arrayLayers = 1u;
+
+			auto& image = ICPUImage::create(std::move(params));
+
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(image->getImageDataSizeInBytes());
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+			ICPUImage::SBufferCopy& region = regions->front();
+			//region.imageSubresource.aspectMask = ...; // waits for Vulkan
+			region.imageSubresource.mipLevel = 0u;
+			region.imageSubresource.baseArrayLayer = 0u;
+			region.imageSubresource.layerCount = 1u;
+			region.bufferOffset = 0u;
+			region.bufferRowLength = calcPitchInBlocks(width, getTexelOrBlockBytesize(image->getCreationParameters().format));
+			region.bufferImageHeight = 0u;
+			region.imageOffset = { 0u, 0u, 0u };
+			region.imageExtent = image->getCreationParameters().extent;
+
+			float* dataToSend = _IRR_NEW_ARRAY(float, image->getImageDataSizeInBytes());
+			for (uint32_t i = 0; i < image->getImageDataSizeInBytes(); i += 4)
+			{
+				dataToSend[i] = pixels[i]->r;
+				dataToSend[i + 1] = pixels[i]->g;
+				dataToSend[i + 2] = pixels[i]->b;
+				dataToSend[i + 3] = pixels[i]->a;
```

Comment:
doesn't OpenEXR have a function other than `readRgba` that will give you quantized `uint16_t/halfFloat` directly instead of dequantizing into full floats?

Also your image format you declared as RGBA half float above, your buffers are also same size as a half float image, and here you're reading full floats and then memcpying them to buffer. this will be a disaster.

Reply:
You need to know when file is `ImfRgba` (half) and you need `EF_R16G16B16A16_SFLOAT`
https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfCRgbaFile.h#L73

And when we have `Rgba` (full float)  and you need `EF_R32G32B32A32_SFLOAT`
https://github.com/AcademySoftwareFoundation/openexr/blob/625b95fa026c3b78e537e9bb6a39fcd51920ad13/OpenEXR/IlmImf/ImfRgba.h#L55

I don't know if OpenEXR allows for mixed precision channels (R Half, G full) if you find out that it does, then convert to highest one present (if one channel is 32bit, then all 32bit).

Reply:
As I can see they call RGBA something that uses their `Half`s being 16 bit precision floats

Reply:
see the struct declarations/definitions, there's one for full floats and one for half.

Reply:
https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfCRgbaFile.h#L50 `IlmHalf`
and `Half` https://github.com/AcademySoftwareFoundation/openexr/blob/625b95fa026c3b78e537e9bb6a39fcd51920ad13/IlmBase/Half/half.h#L41 right?

Reply:
they're all stored as uint16_t, its just that the bits have a different meaning than for an actual `uint16_t`

Reply:
well, because of that they provide this though, it will be good for me https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfConvert.h#L75

Reply:
you dont' want to convert

an array of IlmHalf is good to do a straight copy to a buffer for an EF_R16G16B16A16_SFLOAT texture